### PR TITLE
jit: slice the kind in rewrite_op_getarrayitem so the raw_r guard can fire

### PR DESCRIPTION
**Pull request by youknowone**
Unify indexing to `getkind(...)[0]` to avoid manual check for every use case

### CODE REVIEW

File: rpython/jit/codewriter/jtransform.py [hunk 1]
```diff
@@ -771,15 +771,15 @@ def rewrite_op_getarrayitem(self, op):
         if immut:
             pure = '_pure'
         arraydescr = self.cpu.arraydescrof(ARRAY)
-        kind = getkind(op.result.concretetype)
+        kind = getkind(op.result.concretetype)[0]
         if ARRAY._gckind != 'gc':
             assert ARRAY._gckind == 'raw'
             if kind == 'r':
```
Comment (youknowone) [inline rpython/jit/codewriter/jtransform.py:777]: this line `kind == 'r'` was comparing `getkind(...)` to `r` instead of `getkind(...)[0]` like other usages

**Comment by cfbolz** [comment]
Again, I would be interested in knowing whether there is a wider effect of this. Eg can you write a test metainterp/test/test_ajit.py that gets fixed by this change?

**Comment by youknowone** [comment]
@cfbolz I didn't find this by real code, so not sure it affect to real pypy production. 

~~Added a generated test but didn't check it yet. I will check it later today~~

File: rpython/jit/metainterp/test/test_ajit.py [hunk 2]
```diff
@@ -3255,6 +3255,21 @@ def f1(n):
...
+        # letting the blackhole interpreter fail to find a handler for it.
+        S = lltype.GcStruct('S', ('x', lltype.Signed))
+        A = rffi.CArray(lltype.Ptr(S))
+        def f(n):
+            a = lltype.malloc(A, 5, flavor='raw', zero=True)
+            res = 1 if a[n] else 0
+            lltype.free(a, flavor='raw')
+            return res
+        e = py.test.raises(Exception, self.interp_operations, f, [0])
+        assert 'getarrayitem_raw_r not supported' in str(e.value)
```
Comment (youknowone) [inline rpython/jit/metainterp/test/test_ajit.py:3271]: Before the patch, it raises `AttributeError: type object 'BlackholeInterpreter' has no attribute 'bhimpl_getarrayitem_raw_r'` in blackhole.py

```
Traceback (most recent call last):
  File "p4_ajit_probe.py", line 31, in <module>
    res = r.interp_operations(f, [0])
  File "rpython/jit/metainterp/test/support.py", line 273, in interp_operations
    result1 = _run_with_blackhole(self, args)
  File "rpython/jit/metainterp/test/support.py", line 113, in _run_with_blackhole
    blackholeinterpbuilder = BlackholeInterpBuilder(cw)
  File "rpython/jit/metainterp/blackhole.py", line 58, in __init__
    self.setup_insns(asm.insns)
  File "rpython/jit/metainterp/blackhole.py", line 80, in setup_insns
    all_funcs.append(self._get_method(name, argcodes))
  File "rpython/jit/metainterp/blackhole.py", line 238, in _get_method
    unboundmethod = getattr(BlackholeInterpreter, 'bhimpl_' + name).im_func
AttributeError: type object 'BlackholeInterpreter' has no attribute 'bhimpl_getarrayitem_raw_r'
```

**Comment by youknowone** [comment]
For more details.

I am working for https://github.com/youknowone/pyre, and because it requires agents to read PyPy code a lot, it occasionally find PyPy bugs and suspicious code smells. Sometimes it is a production glitch (#5508) and sometimes potential risk like this. I actually got more reports about smells but didn't submit them because I wasn't sure they are real bugs.
To be honest, I don't understand pypy codebase well to understand the reports enough and patch myself. I try to confirm they are real issues. I am not sure the patch is correct, but usually sure the report is real.
@cfbolz if you don't like patches like this, not real bugs observable in pypy, please feel free to tell me. I can stop them and submit only real production bug fixes. Otherwise I will try to append enough test and explanation, but unlike the real bugs, tests and explanation about them will be mostly generated because I don't understand them that much.

**Comment by cfbolz** [comment]
No, I like your patches! They seem to find real corner cases. I just want to 1) think about them for a few days because I'm worried about unintended consequences and 2) make sure that the right kind of test is also added.
